src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+	 */
+	if (IsA(node, List))
+	{
+		/* at the top of quals we'll frequently see lists, those are to be treated as ANDs */
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *boolExpr = (BoolExpr *) node;
+
+		if (boolExpr->boolop == NOT_EXPR)
+		{
+			return false;
+		}
+		else if (boolExpr->boolop == AND_EXPR)
+		{
+			return expression_tree_walker((Node *) boolExpr->args,
+										  PrunableExpressionsWalker, context);
+		}
+		else if (boolExpr->boolop == OR_EXPR)
+		{
+			ListCell *opCell = NULL;
+
+			/*
+			 * "Queue" partial pruning instances. That's done so they'll get
+			 * any other unconditional ANDed expressions.  Then after that the
+			 * queued expressions will be further processed in
+			 * PrunableExpressions().
+			 */
+			foreach(opCell, boolExpr->args)
+			{
+				PendingPruningInstance *instance =
+					palloc0(sizeof(PendingPruningInstance));
+
+				instance->instance = context->currentPruningInstance;
+				instance->continueAt = lfirst(opCell);
+
+				/*
+				 * Signal that this instance is not to be used for pruning on
+				 * its own.  Once the pending instance is processed, it'll be
+				 * used.
+				 */
+				instance->instance->isPartial = true;
+
+				context->pendingInstances = lappend(context->pendingInstances, instance);
+			}
+
+			return false;
+		}
+	}
+	else if (IsA(node, OpExpr))
+	{
+		OpExpr *operatorExpression = (OpExpr *) node;
+		PruningInstance *prune = context->currentPruningInstance;
+		Node *leftOperand = NULL;
+		Node *rightOperand = NULL;
+		Const *constantClause = NULL;
+		Var *varClause = NULL;
+
+		if (!prune->added)
+		{
+			context->pruningInstances = lappend(context->pruningInstances,
+												prune);
+			prune->added = true;
+		}
+
+		if (list_length(operatorExpression->args) == 2)
+		{
+			leftOperand = get_leftop((Expr *) operatorExpression);
+			rightOperand = get_rightop((Expr *) operatorExpression);
+
+			leftOperand = strip_implicit_coercions(leftOperand);
+			rightOperand = strip_implicit_coercions(rightOperand);
+
+			if (IsA(rightOperand, Const) && IsA(leftOperand, Var))
+			{
+				constantClause = (Const *) rightOperand;
+				varClause = (Var *) leftOperand;
+			}
+			else if (IsA(leftOperand, Const) && IsA(rightOperand, Var))
+			{
+				constantClause = (Const *) leftOperand;
+				varClause = (Var *) rightOperand;
+			}
+		}
+
+		if (constantClause && varClause && equal(varClause, context->partitionColumn))
+		{
+			List *btreeIntepretationList = NULL;
+			ListCell *btreeInterpretationCell = NULL;
+			bool matchedOp = false;
+
+			/*
+			 * Found a restriction on the partition column itself. Update the
+			 * current constraint with the new information.
+			 */
+
+			btreeIntepretationList =
+				get_op_btree_interpretation(operatorExpression->opno);
+			foreach(btreeInterpretationCell, btreeIntepretationList)
+			{
+				OpBtreeInterpretation *btreeIntepretation =
+					(OpBtreeInterpretation *) lfirst(btreeInterpretationCell);
+
+				switch (btreeIntepretation->strategy)
```

Comment:
I tried this, and it imo doesn't seem to come out easier to understand.  Unless you feel strongly, I'd rather not go for it.